homeassistant/components/tankerkoenig/__init__.py

```diff
@@ -1,174 +1,202 @@
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload Tankerkoenig config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.unique_id)
+
+    return unload_ok
+
+
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+class TankerkoenigDataUpdateCoordinator(DataUpdateCoordinator):
     """Get the latest data from the API."""
 
-    def __init__(self, hass, conf):
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        logger: logging.Logger,
+        name: str,
+        update_interval: int,
+    ) -> None:
         """Initialize the data object."""
-        self._api_key = conf[CONF_API_KEY]
-        self.stations = {}
-        self.fuel_types = conf[CONF_FUEL_TYPES]
-        self.update_interval = conf[CONF_SCAN_INTERVAL]
-        self.show_on_map = conf[CONF_SHOW_ON_MAP]
-        self._hass = hass
-
-    def setup(self, latitude, longitude, radius, additional_stations):
-        """Set up the tankerkoenig API.
 
-        Read the initial data from the server, to initialize the list of fuel stations to monitor.
-        """
-        _LOGGER.debug("Fetching data for (%s, %s) rad: %s", latitude, longitude, radius)
-        try:
-            data = pytankerkoenig.getNearbyStations(
-                self._api_key, latitude, longitude, radius, "all", "dist"
-            )
-        except pytankerkoenig.customException as err:
-            data = {"ok": False, "message": err, "exception": True}
-        _LOGGER.debug("Received data: %s", data)
-        if not data["ok"]:
-            _LOGGER.error(
-                "Setup for sensors was unsuccessful. Error occurred while fetching data from tankerkoenig.de: %s",
-                data["message"],
-            )
-            return False
+        super().__init__(
+            hass=hass,
+            logger=logger,
+            name=name,
+            update_interval=timedelta(minutes=update_interval),
+        )
 
-        # Add stations found via location + radius
-        if not (nearby_stations := data["stations"]):
-            if not additional_stations:
-                _LOGGER.error(
-                    "Could not find any station in range."
-                    "Try with a bigger radius or manually specify stations in additional_stations"
-                )
-                return False
-            _LOGGER.warning(
-                "Could not find any station in range. Will only use manually specified stations"
-            )
-        else:
-            for station in nearby_stations:
-                self.add_station(station)
+        self._api_key = entry.data[CONF_API_KEY]
+        self._selected_stations = entry.data[CONF_STATIONS]
+        self._hass = hass
```

Comment:
We don't need this. `hass` is available here:
https://github.com/home-assistant/core/blob/d59ecc4c963ffbeb80c626de6c600170ed85e165/homeassistant/helpers/update_coordinator.py#L50